### PR TITLE
Agolubev #18890 polish source.queue

### PR DESCRIPTION
**Pull request by agolubev**
Ref: #18890 and this is previous PR: #19296

### CODE REVIEW

**Comment by drewhk** [comment]
OK TO TEST

**Comment by agolubev** [comment]
tests failed
```
[info] - should catch response stream truncation *** FAILED *** (28 milliseconds)
[info]   Expected exception akka.stream.scaladsl.One2OneBidiFlow$OutputTruncationException$ to be thrown, but java.lang.IllegalArgumentException was thrown. (HighLevelOutgoingConnectionSpec.scala:70)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:496)
[info]   at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1538)
[info]   at org.scalatest.Assertions$.checkExpectedException(Assertions.scala:1572)
[info]   at org.scalatest.words.ResultOfBeWordForAType.thrownBy(ResultOfBeWordForAType.scala:39)
[info]   at akka.http.impl.engine.client.HighLevelOutgoingConnectionSpec$$anonfun$1$$anonfun$apply$mcV$sp$3.apply$mcV$sp(HighLevelOutgoingConnectionSpec.scala:83)
```

```
[info] - must #toStrict with failing source *** FAILED *** (47 milliseconds)
[info]   "toStrict received OnError while draining stream! Accumulated elements: List(1, 2)" did not include substring "Boom!" (StreamTestKitSpec.scala:40)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.MatchersHelper$.newTestFailedException(MatchersHelper.scala:160)
[info]   at org.scalatest.Matchers$ShouldMethodHelper$.shouldMatcher(Matchers.scala:6231)
[info]   at org.scalatest.Matchers$AnyShouldWrapper.should(Matchers.scala:6265)
[info]   at akka.stream.testkit.StreamTestKitSpec$$anonfun$1$$anonfun$apply$mcV$sp$2.apply$mcV$sp(StreamTestKitSpec.scala:40)
[info]   at akka.stream.testkit.StreamTestKitSpec$$anonfun$1$$anonfun$apply$mcV$sp$2.apply(StreamTestKitSpec.scala:24)
[info]   at akka.stream.testkit.StreamTestKitSpec$$anonfun$1$$anonfun$apply$mcV$sp$2.apply(StreamTestKitSpec.scala:24)
```

```
[info] - must timeout and cancel substream publishers when no-one subscribes to them after some time (time them out) *** FAILED *** (1 second, 577 milliseconds)
[info]   3 did not equal "expected" (SubstreamSubscriptionTimeoutSpec.scala:78)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.MatchersHelper$.newTestFailedException(MatchersHelper.scala:160)
[info]   at org.scalatest.Matchers$ShouldMethodHelper$.shouldMatcher(Matchers.scala:6231)
[info]   at org.scalatest.Matchers$AnyShouldWrapper.should(Matchers.scala:6277)
[info]   at akka.stream.scaladsl.SubstreamSubscriptionTimeoutSpec$$anonfun$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$5.apply(SubstreamSubscriptionTimeoutSpec.scala:78)
[info]   at akka.stream.scaladsl.SubstreamSubscriptionTimeoutSpec$$anonfun$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$5.apply(SubstreamSubscriptionTimeoutSpec.scala:43)
```

File: akka-stream/src/main/scala/akka/stream/Queue.scala [hunk 1]
```diff
@@ -12,13 +12,21 @@ trait SourceQueue[T] {
 
   /**
    * Method offers next element to a stream and returns future that:
-   * - competes with true if element is consumed by a stream
-   * - competes with false when stream dropped offered element
-   * - fails if stream is completed or cancelled.
+   * - completes with `Success(true)` if element is consumed by a stream
+   * - completes with `Success(false)` when stream dropped offered element
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/Queue.scala]: I would rather have a `Dropped` status instead.

File: akka-stream/src/main/scala/akka/stream/Queue.scala [hunk 2]
```diff
@@ -12,13 +12,21 @@ trait SourceQueue[T] {
 
   /**
    * Method offers next element to a stream and returns future that:
-   * - competes with true if element is consumed by a stream
-   * - competes with false when stream dropped offered element
-   * - fails if stream is completed or cancelled.
+   * - completes with `Success(true)` if element is consumed by a stream
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/Queue.scala]: below `Success[T]` has a type parameter `T` but no boolean field.

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/Queue.scala]: ah nevermind I see you parametrized here with a Boolean.

File: akka-stream/src/main/scala/akka/stream/Queue.scala [hunk 3]
```diff
@@ -35,3 +43,29 @@ trait SinkQueue[T] {
    */
   def pull(): Future[Option[T]]
 }
+
+sealed abstract class StreamCallbackStatus[+T]
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/Queue.scala]: why is the type parameter needed?

File: akka-stream/src/main/scala/akka/stream/Queue.scala [hunk 4]
```diff
@@ -35,3 +43,29 @@ trait SinkQueue[T] {
...
+
+/**
+ * Contains types that is used as return types for async callbacks to streams
+ */
+object StreamCallbackStatus {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/Queue.scala]: Hm, I am not sure if this needs to be so general. You can just have `QueueOfferResult` with various outcomes:
 - Enqueued
 - Dropped
 - ProtocolViolation("cannot do x while y")
 - QueueClosed

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 5]
```diff
@@ -1342,3 +1340,55 @@ abstract class AbstractOutHandler extends OutHandler
...
+  private trait CallbackState
+  private case class NotInitialized(list: List[T]) extends CallbackState
+  private case class Initialized(f: T ⇒ Unit) extends CallbackState
+  private case class Stopped(f: T ⇒ Unit) extends CallbackState
+
+  /*
+   * To preserve message order when switching between not initialized / initialized states
+   * lock is used. Case is similar to RepointableActorRef
+   */
+  private[this] final val lock = new ReentrantLock
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: argh, a lock. @rkuhn I recall that you have commented something regarding this, but is a lock really needed? I don't remember what was the original issue here, but this looks dangerous.

File: akka-stream/src/main/scala/akka/stream/Queue.scala (hunk 4 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/Queue.scala]: I actually tried to make it compliant with SinkQueue#pull. Where we can do get result as T if steam is has something

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala (hunk 5 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: https://github.com/akka/akka/pull/19296/files#r49307808

File: akka-stream/src/main/scala/akka/stream/Queue.scala (hunk 4 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/Queue.scala]: I am not sure this helps in this case given that you are unlikely reuse enqueueing logic as dequeueing logic and vice versa, which means that there is no real use case to make this generic (apart from having one less class, but I think more meaningful, domain specific return types are worth the effort here)

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala (hunk 5 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: >while traversing this list the client could concurrently already add new callbacks that go directly to the GraphStage, bypassing this queue and thereby mixing up the ordering. Adding a lock here and around the Initialized case of callback below would fix it, but we hate that solution.

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: well, the " we hate that solution." part is why I asked. I don't know if this issue was fully decided on or not, so I asked @rkuhn .

File: akka-stream/src/main/scala/akka/stream/Queue.scala (hunk 4 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/Queue.scala]: Ok, makes sense.

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala (hunk 5 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: «We hates it, but we sees no other solutionses …»

Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: _(gollum)_

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Ok.

**Comment by rkuhn** [comment]
needs a rebase and validation and then it can be merged

**Comment by rkuhn** [comment]
(I can do that if you don’t have time today, @agolubev )

**Comment by agolubev** [comment]
rkuhn, I can do it during night hours and push prior your morning tomorrow. If you need it earlier, please jump in.

**Comment by rkuhn** [comment]
Okay, will first finish the other things and notify this ticket when starting to work on it—then we’ll see who gets there faster :-) (please also ping when starting to work on it).

**Comment by agolubev** [comment]
Ok, will do

**Comment by agolubev** [comment]
Ok, almost done with this

**Comment by agolubev** [comment]
@rkuhn, just pushed fixes according to last fedback

**Comment by agolubev** [comment]
failed with unrelated reason:
```
[info] - should not open a second connection if there is an idle one available *** FAILED *** (3 seconds, 70 milliseconds)
[info]   java.lang.AssertionError: Expected OnNext(_), yet no element signaled during 3 seconds
[info]   at akka.stream.testkit.TestSubscriber$ManualProbe.expectNext(StreamTestKit.scala:257)
[info]   at akka.http.impl.engine.client.ConnectionPoolSpec$$anonfun$1$$anonfun$apply$mcV$sp$4$$anon$7.<init>(ConnectionPoolSpec.scala:126)
[info]   at akka.http.impl.engine.client.ConnectionPoolSpec$$anonfun$1$$anonfun$apply$mcV$sp$4.apply$mcV$sp(ConnectionPoolSpec.scala:115)
```

**Comment by rkuhn** [comment]
Great work again @agolubev !

**Pull request merged**